compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -787,6 +790,17 @@ static LogicalResult setMatmulPeelingRootConfig(
     clampedVecTileSizes[index] = std::min(clampedVecTileSizes[index], size);
   }
 
+  // The LLVM backend struggles to legalize non-power-of-two scalable vectors,
+  // hence the extra rounding up.
```

Comment:
I would say this is true for all the backends. Do we have cases where we generate non power of two for the main vector dimension? 

We also have to make sure we don't round up sizes of unrolled dimensions. It should be fine to unroll by 7, for example.

Reply:
I wonder that if this makes sense when the dim size is smaller than vector size? @banach-space mentioned that the "clamping" is needed in scalable vectorization. I think he is evaluating if it is really needed or not.

Reply:
I think this was originally my comment, what I meant here was fixed vectors can always be scalarized (though may generate poor code). Scalable vectors cannot always be scalarized/legalized, so can result in the backend crashing. E.g.  `vector<[7]xf32>`, probably can't be legalized.

Reply:
>  Do we have cases where we generate non power of two for the main vector dimension?
It did used to happen for statically sized tensors.

Reply:
> Do we have cases where we generate non power of two for the main vector dimension?

In the past I've noticed that happening due to "clamping" - that's one reason I'd like to remove it.

> I wonder that if this makes sense when the dim size is smaller than vector size? @banach-space mentioned that the "clamping" is needed in scalable vectorization. I think he is evaluating if it is really needed or not.

I will prioritise that next week 👍🏻

Reply:
I think this is a legacy decision that we made in an attempt to make the inner tile sizes multiple of the distribution sizes but I think it doesn't hold anymore. When we use masking, we want to always round up to the nearest power of two. However, when we use peeling, we want to make sure the main vector loop executes at least once so we should round up to the nearest power of two only if there are enough iterations for that to happen (or an unknown number of iterations). Otherwise, we should round down to the nearest power of two and let the remainder kick in.